apps/prairielearn/elements/pl-file-upload/pl-file-upload.py

```diff
@@ -86,41 +239,93 @@ def render(element_html: str, data: pl.QuestionData) -> str:
-        if x.get("name", "") in required_file_names:
-            pl.add_submitted_file(data, x.get("name", ""), x.get("contents", ""))
-
-    # Validate that all required files are present
     if parsed_files is not None:
-        submitted_file_names = [x.get("name", "") for x in parsed_files]
-        missing_files = [
-            x for x in required_file_names if x not in submitted_file_names
-        ]
+        # Files with HTML characters "<" or ">" should have been rejected client-side already, so we silently filter them here
+        parsed_file_names = [f.get("name", "") for f in parsed_files]
+
+        # Split optional names into two separate lists: wildcard patterns and plain names
+        files_regex = [glob_to_regex(f) for f in file_patterns]
+        opt_files_regex = [glob_to_regex(f) for f in opt_file_patterns]
+
+        # Match submitted and accepted files 1:1 in this order: required_files, then required patterns, then optional files.
+        # All remaining files are then matched 1:n with optional patterns, and eventually discarded in there is no match
+        required_files = [x for x in parsed_file_names if x in file_names]
+        remaining_files = [x for x in parsed_file_names if x not in required_files]
+
+        pattern_files, missing_regex = match_regex_with_files(
+            files_regex,
+            [x for x in remaining_files],
+            limit_1=True,
+        )
+        remaining_files = [x for x in remaining_files if x not in pattern_files]
+
+        # Validate that all required files are present
+        missing_files = [x for x in file_names if x not in parsed_file_names]
 
         if len(missing_files) > 0:
             add_format_error(
+                answer_name,
+                data,
+                f"The following required files were missing: {', '.join(missing_files)}",
+            )
+
+        if len(missing_regex) > 0:
+            # Need some awkward mapping to patterns here to ensure we don't remove duplicates
+            missing_patterns = []
+            for r in missing_regex:
+                missing_patterns.append(file_patterns[files_regex.index(r)])
+            add_format_error(
+                answer_name,
                 data,
-                "The following required files were missing: "
-                + ", ".join(missing_files),
+                f"The following required file patterns were missing: {', '.join(missing_patterns)}",
             )
+
+        optional_files = [x for x in remaining_files if x in opt_file_names]
+        remaining_files = [x for x in remaining_files if x not in optional_files]
+        # We don't care which optional patterns are matched
+        opt_pattern_files, _ = match_regex_with_files(
+            opt_files_regex,
+            [x for x in remaining_files],
+            limit_1=False,
+        )
+
+        # Finally, we combine all file categories into one big list
+        parsed_files = [
```

Comment:
Maybe put all of these lists into a set for the purposes of doing this filtering? This is a little hard to read. Instead of a list, you could also combine this with the for loop below and skip the `add_submitted_file` call if the obtained name is in this set, like:
```python
include_set = required_files | pattern_files | optional_files | opt_pattern_files
for file in parsed_files:
    file_name = file.get("name", "")
    if file_name in include_set:
        pl.add_submitted_file(data, file_name, file.get("contents", ""))
```